SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/PredictedGameplayCuesTest/PredictedGameplayCuesTest.cpp

```diff
@@ -0,0 +1,172 @@
+		// (by default it searches the game content folder for assets).
+		// If a cue is not registered with the manager, cue events will never be routed to it.
+		// That means we have to register the cue manually here on each client.
+		FSoftObjectPath CuePath(TEXT("/Script/SpatialGDKFunctionalTests.GC_SignalCueActivation"));
+		FGameplayCueReferencePair ExecuteCueRef(UGC_SignalCueActivation::GetExecuteTag(), CuePath);
+		FGameplayCueReferencePair AddCueRef(UGC_SignalCueActivation::GetAddTag(), CuePath);
+		UAbilitySystemGlobals::Get().GetGameplayCueManager()->GetRuntimeCueSet()->AddCues({ ExecuteCueRef, AddCueRef });
+
+		FinishStep();
+	});
+
+	auto WaitForEventConfirmation = [this](int Counter, float DeltaTime) {
+		if (bTimerStarted)
+		{
+			StepTimer += DeltaTime;
+		}
+
+		if (Counter == 0)
+		{
+			// The counter is allowed to be 0 while we are still waiting for the changed counter to replicated to us.
+			// However, if we've seen Counter at 1 before (and as a result started the timer), if we see it back at 0,
+			// it somehow got reset. This likely indicates a bug in the test implementation.
+			if (bTimerStarted)
+			{
+				FinishTest(EFunctionalTestResult::Invalid, TEXT("The counter was reset to 0 after being set to 1."));
+			}
+		}
+		else if (Counter == 1)
+		{
+			if (!bTimerStarted)
+			{
+				bTimerStarted = true;
+			}
+			else if (StepTimer >= DuplicateActivationCheckWaitTime)
+			{
+				bTimerStarted = false;
+				StepTimer = 0.0f;
+				FinishStep();
+			}
+		}
+		else
+		{
+			FinishTest(EFunctionalTestResult::Failed,
+					   FString::Printf(TEXT("The event was triggered %d times. It should only be triggered once."), Counter));
+		}
+	};
+
+	AddStep(
+		TEXT("Spawn Actor"), FWorkerDefinition::Server(1), nullptr,
+		[this]() {
+			UWorld* World = GetWorld();
+			// The exact spawn location doesn't matter much. 10,10 should at least not be on the boundary between servers
+			TestPawn = World->SpawnActor<ACuesGASTestPawn>({ 10.0f, 10.0f, 0.0f }, FRotator::ZeroRotator);
+			AssertTrue(IsValid(TestPawn), TEXT("Target actor is valid."));
+			AssertEqual_Int(TestPawn->GetOnActiveCounter(), 0, TEXT("OnActive counter should start at 0."));
+			AssertEqual_Int(TestPawn->GetExecuteCounter(), 0, TEXT("Executed counter should start at 0."));
+
+			// Set the target to be owned by client 1. An actor has to be client-owned to be able to run client-predicted gameplay abilities
+			// on it.
+			ASpatialFunctionalTestFlowController* FlowController = GetFlowController(ESpatialFunctionalTestWorkerType::Client, 1);
+			AController* PlayerController = static_cast<AController*>(FlowController->GetOwner());
+
+			// Store the current pawn so we can re-possess it during cleanup
+			PrevPawn = PlayerController->GetPawn();
+			PlayerController->Possess(TestPawn);
+
+			RegisterAutoDestroyActor(TestPawn);
+		},
+		[this](float DeltaTime) {
+			if (TestPawn->IsActorReady())
```

Comment:
could use RequirTrue here instead

Reply:
I don't think so, RequireTrue would just fail the step as soon as `IsActorReady()` is false. The point of this is to wait until it becomes ready, or else never finish the step and thus time out.

Reply:
That is exactly what Require does, Assert will fail immediately but Require will check a condition each tick until the time out is up or the condition is met.

Reply:
Ah my bad, I wasn't aware these existed and just mentally substituted require with assert. Will read up on how exactly these work and substitute appropriately, thanks.